File: 290-word-pattern/word-pattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        String[] words = s.split(" ");

        if(words.length != pattern.length()){
            return false;
        }  

        HashMap <Character,String> map1 = new HashMap<>();
        HashMap <String,Character> map2 = new HashMap<>();

        for(int i=0;i<pattern.length();i++){
            char c = pattern.charAt(i);
            String word = words[i];

        if(map1.containsKey(c)){
            if(!map1.get(c).equals(word)){
                return false;
            }
        }else{
            map1.put(c,word);
        }

        if(map2.containsKey(word)){
            if(map2.get(word) != c){
                return false;
            }
        }else{
            map2.put(word,c);
        }
        }
        return true;
    }
}

File: 290-word-pattern/word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        a=s.split()
        freq={}
        back={}
        if len(pattern) != len(a):
            return False
        for i in range(len(pattern)):
            if pattern[i] in freq:
                if freq[pattern[i]] != a[i]:
                    return False
            else:
                freq[pattern[i]] = a[i]
            if a[i] in back:
                if back[a[i]] != pattern[i]:
                    return False
            else:
                back[a[i]] = pattern[i]            
        return True                 
        